src/utils/iter.rs

```diff
@@ -0,0 +1,71 @@
+/// Common iterator types
+use wayland_server::{backend::ClientId, Resource, Weak};
+
+use std::{fmt, sync::MutexGuard};
+
+/// Iterator helper over a mutex of client objects
+pub struct LockedClientObjsIter<'a, T: 'static, G, F> {
+    pub iterator: std::iter::FilterMap<std::slice::Iter<'static, Weak<T>>, F>,
+    pub guard: MutexGuard<'a, G>,
+}
```

Comment:
So, `iterator` includes a borrow into `guard`? Is this sound under aliasing rules?

Reply:
(And presumably the fields shouldn't be `pub` here.)

Reply:
Yes, because it actually has a pointer into the Mutex, which is inside the guard behind a reference. And because we don't modify the guard in any way the compiler won't allow moving the original mutex while it is borrowed.

Reply:
> (And presumably the fields shouldn't be `pub` here.)

Doesn't make a difference due to this not being part of the public api, but good catch.

Reply:
Right, I guess the fields aren't exposed since this is only used in the API as `Into<Iterator>` (so, both the fields and the struct itself don't need to be `pub`.)

Yeah, I guess this isn't an aliasing issue, looking at how libstd defines `Mutex` and `MutexGuard`. Technically this may be an implementation detail (if there's no explicit guarantee that this is safe to do), but should be fine. Hopefully `MutexGuard::map` gets stabilized before *too* long.